openedx/core/lib/xblock_runtime/utils.py

```diff
@@ -0,0 +1,63 @@
+"""
+Useful helper methods related to the XBlock runtime
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+import hashlib
+import hmac
+import logging
+import math
+import time
+
+from django.conf import settings
+from six import text_type
+
+
+def get_secure_token_for_xblock_handler(user_id, block_key_str, time_idx = 0):
+    """
+    Get a secure token (one-way hash) used to authenticate XBlock handler
+    requests. This token replaces both the session ID cookie (or OAuth
+    bearer token) and the CSRF token for such requests.
+
+    The token is specific to one user and one XBlock usage ID, though may
+    be used for any handler. It expires and is only valid for 2-4 days.
+
+    We use this token because XBlocks may sometimes be sandboxed (run in a
+    client-side JavaScript environment with no access to cookies) and
+    because the XBlock python and JavaScript handler_url APIs do not provide
+    any way of authenticating the handler requests, other than assuming
+    cookies are present or including this sort of secure token in the
+    handler URL.
+
+    to ensure that only authorized XBlock code in an IFrame we created is
+    calling the secure_xblock_handler endpoint.
+
+    For security, we need these tokens to have an expiration date. So: the
+    hash incorporates the current time, rounded to the lowest TOKEN_PERIOD
+    value. When checking this, you should check both time_idx=0 and
+    time_idx=-1 in case we just recently moved from one time period to
+    another (i.e. at the stroke of midnight UTC or similar). The effect of
+    this is that each token is valid for 2-4 days.
```

Comment:
Hmm. Not sure if this variability is a good idea.

Reply:
The alternative is to return a different token each time this is called, with an expiry date of exactly 2*86400 seconds from the time of the call. But then you have to store the tokens server-side in order to "remember" when each token expires. I prefer this approach of having a consistent token for repeated calls within the same time window, because it doesn't require storing any data server-side and it's simpler to debug (the token won't change throughout the typical day)

Reply:
Makes sense. May be useful to document this reasoning. 👍

Reply:
Good point. Added to the docstring.